Secondary boot handling

igoropaniuk opened the pull request:
Please do not submit a Pull Request via github.  Our project makes use of
mailing lists for patch submission and review.  For more details please
see https://www.denx.de/wiki/U-Boot/Patches

igoropaniuk committed:
[FIO internal] imx: extend secondary_boot cmd

Add possibility to retrieve current value for the flag.

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

angolini commented on cmd/fiovb.c:
```diff
@@ -142,11 +142,35 @@ int do_fiovb_delete_pvalue(cmd_tbl_t *cmdtp, int flag, int argc,
+
+	if (fiovb_ops->is_secondary_boot(fiovb_ops, &secondary) ==
+	    FIOVB_IO_RESULT_OK) {
+		printf("Secondary bit = %d\n", secondary);
```
it made me think if we need a more verbose return here. I like the other printf line (with the run 'fiovb init' first), and I wonder what happen if the result of this command is only "Secondary bit 1"

anyway, only aesthetic comment, if this is the only one, I would say go for the merge.

ricardosalveti replied:
I wonder if it is better to just not output anything, as I get the impression we're currently way too verbose with fiovb.

This will later be used as part of our boot script, right?

igoropaniuk replied:
@ricardosalveti exactly.
Return code is enough to this cmd in `if else` build-it statements in U-Boot shell
However additional output would be good for debugging purposes (so I can just `s/printf/debug/g` here)

igoropaniuk committed:
[FIO internal] imx8mm_evk: don't autoset secondary boot bit

Don't autoset seconday boot bit, as now we boot into primary boot
path by default on WDOG reset.

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

igoropaniuk committed:
[FIO internal] imx8mm_evk: add verbose output about next boot image

Add verbose about offset used for the next boot image.

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

igoropaniuk committed:
[FIO internal] imx: add imx_warm_reset cmd

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

ricardosalveti commented on include/fiovb.h:
```diff
@@ -50,6 +50,10 @@ struct fiovb_ops {
+
+  fiovb_io_result (*secondary_boot)(struct fiovb_ops* ops,
+                                    int* secondary_boot,
+				    bool update);
```
55 and 56 are mixed with tabs x spaces.

igoropaniuk replied:
yeah, will fix

ricardosalveti commented:
Minor comment, otherwise LGTM.

mike-scott commented on board/freescale/imx8mm_evk/spl.c:
```diff
@@ -291,24 +291,13 @@ unsigned long spl_mmc_get_uboot_raw_sector(struct mmc *mmc,
-
-		debug("%s: Set: GPR10 PERSIST_SECONDARY_BOOT = %d\n",
-		      __func__, boot_mode_getprisec());
+		printf("SPL: Booting secondary boot path: using 0x%lx offset "
```
Should we change these to `debug` ?

igoropaniuk replied:
To be honest I'd prefer to have that information always printed

mike-scott commented on board/freescale/imx8mm_evk/spl.c:
```diff
@@ -291,24 +291,13 @@ unsigned long spl_mmc_get_uboot_raw_sector(struct mmc *mmc,
+		printf("SPL: Booting secondary boot path: using 0x%lx offset "
+		       "for next boot image\n", offset);
+	} else {
+		printf("SPL: Booting primary boot path: using 0x%lx offset "
```
Here as well.

mike-scott reviewed (commented):
Looks very close to me:
- the 1 comment from @ricardosalveti on tabs/spaces
- question on whether we want to print the boot info all the time or use `debug`

ldts commented on cmd/fiovb.c:
```diff
@@ -142,11 +142,69 @@ int do_fiovb_delete_pvalue(cmd_tbl_t *cmdtp, int flag, int argc,
+	if (fiovb_ops->secondary_boot(fiovb_ops, &secondary, false) ==
+	    FIOVB_IO_RESULT_OK) {
+		printf("Secondary boot bit = %d\n", secondary);
+		snprintf(fiovb_val, sizeof(fiovb_val), "%d", secondary);
```
maybe check return value

igoropaniuk replied:
that should be fixed everywhere (including do_fiovb_read_pvalue(), do_fiovb_write_pvalue(), do_fiovb_delete_pvalue() etc)

ldts replied:
yes, but that doesnt make this lesss wrong does it?

ldts commented on cmd/fiovb.c:
```diff
@@ -142,11 +142,69 @@ int do_fiovb_delete_pvalue(cmd_tbl_t *cmdtp, int flag, int argc,
+	    FIOVB_IO_RESULT_OK) {
+		printf("Secondary boot bit = %d\n", secondary);
+		snprintf(fiovb_val, sizeof(fiovb_val), "%d", secondary);
+		env_set("fiovb.is_secondary_boot", fiovb_val);
```
maybe check return value (snprintf and env_set)

ldts commented on cmd/fiovb.c:
```diff
@@ -142,11 +142,69 @@ int do_fiovb_delete_pvalue(cmd_tbl_t *cmdtp, int flag, int argc,
+	if (argc != 2)
+		return CMD_RET_USAGE;
+
+	if (!(secondary == 0 || secondary == 1))
```
what is this check doing?

igoropaniuk replied:
good point, it should be after `secondary = simple_strtoul(argv[1], NULL, 10);`
Curious how that even works :)
I can also drop that check, and just do 
secondary = !!secondary

ldts commented on common/fiovb.c:
```diff
@@ -227,6 +227,36 @@ static fiovb_io_result delete_persistent_value(struct fiovb_ops *ops,
+				      bool update)
+{
+	fiovb_io_result rc;
+	struct tee_param param[1];
```
why does it need to be an array of one element?

igoropaniuk replied:
I can change it to (also addressed your next comment):
```
struct tee_param param = { .attr = TEE_PARAM_ATTR_TYPE_VALUE_INOUT };
```

ldts commented on common/fiovb.c:
```diff
@@ -227,6 +227,36 @@ static fiovb_io_result delete_persistent_value(struct fiovb_ops *ops,
+
+	tee = ((struct fiovb_ops_data *)ops->user_data)->tee;
+
+	memset(param, 0, sizeof(param));
```
why not initialize the structure to zeros instead of memset?

ldts commented on arch/arm/mach-imx/imx_secondary_boot.c:
```diff
@@ -48,3 +48,17 @@ U_BOOT_CMD(
+static int do_warm_reset(cmd_tbl_t *cmdtp, int flag,
+			 int argc, char * const argv[])
+{
+	call_imx_sip(IMX_SIP_WARM_RESET, 0, 0, 0, 0);
```
why dont we check the return value?

igoropaniuk replied:
well, we can do that, but it doesn't really make sense as technically we will never return from this call (because of reset done in TF-A)

ldts replied:
so then no return on the function ?

igoropaniuk committed:
[FIO internal] ARM: imx8m: add support for reading out SEC_CONFIG[1]

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

igoropaniuk committed:
[FIO internal] imx: add cmd fo reading out SEC_CONFIG[1]

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

ricardosalveti reviewed (approved):
LGTM, let's merge this and get some testing on it.
